src/System.Net.Primitives/src/System/Net/CookieContainer.cs

```diff
@@ -0,0 +1,1062 @@
+        {
+            lock (source)
+            {
+                // Cannot use foreach as we going update 'source'
+                for (int idx = 0; idx < source.Count; ++idx)
+                {
+                    bool to_add = false;
+
+                    Cookie cookie = source[idx];
+
+                    if (cookie.Expired)
+                    {
+                        // If expired, remove from container and don't add to the destination
+                        source.RemoveAt(idx);
+                        --_count;
+                        --idx;
+                    }
+                    else
+                    {
+                        // Add only if port does match to this request URI
+                        // or was not present in the original response.
+                        if (isPlainOnly && cookie.Variant != CookieVariant.Plain)
+                        {
+                            ; // Don't add
+                        }
+                        else if (cookie.PortList != null)
+                        {
+                            foreach (int p in cookie.PortList)
+                            {
+                                if (p == port)
+                                {
+                                    to_add = true;
+                                    break;
+                                }
+                            }
+                        }
+                        else
+                        {
+                            // It was implicit Port, always OK to add.
+                            to_add = true;
+                        }
+
+                        // Refuse to add a secure cookie into an 'unsecure' destination
+                        if (cookie.Secure && !isSecure)
+                        {
+                            to_add = false;
+                        }
+
+                        if (to_add)
+                        {
+                            // In 'source' are already orederd.
+                            // If two same cookies come from different 'source' then they
+                            // will follow (not replace) each other.
+                            destination.InternalAdd(cookie, false);
+                        }
+                    }
+                }
+            }
+        }
+
+        public string GetCookieHeader(Uri uri)
+        {
+            if (uri == null)
+            {
+                throw new ArgumentNullException("uri");
+            }
+            string dummy;
+            return GetCookieHeader(uri, out dummy);
+        }
+
+        internal string GetCookieHeader(Uri uri, out string optCookie2)
+        {
+            CookieCollection cookies = InternalGetCookies(uri);
+            string cookieString = String.Empty;
+            string delimiter = String.Empty;
+
+            foreach (Cookie cookie in cookies)
```

Comment:
CookieCollection doesn't expose a struct-based, strongly-typed GetEnumerator, so every iteration through it will at least allocate an enumerator object.  Consider changing these to be for loops or adding a such a struct-based enumerator for internal usage (it would probably need some additional boilerplate since there's already a public GetEnumerator that returns IEnumerable).